File: path-sum/path_sum.go
package pathSum

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func hasPathSum(root *TreeNode, targetSum int) bool {
	return hasPath(root, targetSum)
}

func hasPath(node *TreeNode, sum int) bool {
	if node == nil {
		return false
	}

	if isLeafNode(node) {
		if sum-node.Val == 0 {
			return true
		}
		return false
	}

	nextSum := sum - node.Val

	resL := hasPath(node.Left, nextSum)
	if resL {
		return true
	}
	resR := hasPath(node.Right, nextSum)
	if resR {
		return true
	}

	return false
}

func isLeafNode(node *TreeNode) bool {
	if node == nil {
		return false
	}
	if node.Left == nil && node.Right == nil {
		return true
	}

	return false
}

File: tree/path-sum/python/path_sum.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def hasPathSum(self, root: TreeNode, targetSum: int) -> bool:
        if root is None:
            return False
        if self._isLeaf(root):
            if targetSum-root.val == 0:
                return True
            return False

        nextTS = targetSum - root.val

        leftR = self.hasPathSum(root.left, nextTS)
        rightR = self.hasPathSum(root.right, nextTS)

        return leftR | rightR


    def _isLeaf(self, node: TreeNode) -> bool:
        if node is None:
            return False

        if node.left is not None or node.right is not None:
            return False

        return True
